org.eclipse.xtext/src/org/eclipse/xtext/serializer/analysis/GrammarConstraintProvider.java

```diff
@@ -80,36 +83,183 @@ public Constraint(Grammar grammar, EClass type, Nfa<ISemState> nfa) {
+					// The cost of proceeding to this follower is 1 iff it has the correct feature id.
+					int increment = follower.getFeatureID() == featureId ? 1 : 0;
+					if (fdn == null) {
+						// We haven't reached this node before. Assign a new id and distance and mark as unvisited.
+						fdn = new DijkstraNode();
+						fdn.id = nextStateId++;
+						fdn.distance = currentDistance + increment;
+						idAndDistance.put(follower, fdn);
+						unvisited.add(follower);
+					} else {
+						// This follower node has already been reached.
+						fdn.distance = Math.min(fdn.distance, currentDistance + increment);
+					}
+				}
+				unvisited.remove(currentNode);
+				// Choose the unvisited node with the lowest tentative distance.
+				currentNode = unvisited.pollFirst();
+			} while (currentNode != null);
+			// If we get to this point, the stop state is not reachable from the start.
+			throw new AssertionError("Stop state is not reachable.");
+		}
+		
+		/**
+		 * Computing the upper bounds is equivalent to the longest path problem. While this is NP-hard if
+		 * constrained to simple paths, it becomes easily solvable in our case since we include infinite paths.
+		 * All features that are assigned inside a cycle have infinity as upper bound. For the remaining features
+		 * we can handle each cycle as if it was a single node because it is not relevant where we enter the
+		 * cycle and where we exit it (traversing the cycle has zero weight). The resulting simplified graph
+		 * is acyclic, and in such graphs the longest path can be computed in linear time with a simple recursive
+		 * algorithm.
+		 */
+		protected int[] computeUpperBounds() {
+			// A cycle is represented by the set of all nodes on that cycle.
+			Map<ISemState, Set<ISemState>> cycles = Maps.newLinkedHashMap();
+			findCycles(nfa.getStart(), cycles, Maps.newHashMap(), Lists.newLinkedList());
+			
+			// Assign an infinite upper bound to all features that are on a cycle.
+			int[] bounds = new int[type.getFeatureCount()];
+			for (Set<ISemState> cycle : cycles.values()) {
+				for (ISemState node : cycle) {
+					int featureId = node.getFeatureID();
+					if (featureId >= 0) {
+						bounds[featureId] = IGrammarConstraintProvider.MAX;
+					}
+				}
+			}
+			
+			// Handle the remaining features with the linear-time longest path algorithm.
+			for (int i = 0; i < bounds.length; i++) {
+				if (bounds[i] != IGrammarConstraintProvider.MAX) {
+					bounds[i] = computeUpperBound(i, nfa.getStart(), cycles, Maps.newHashMap());
 				}
-				return;
 			}
-			for (ISemState follower : state.getFollowers()) {
-				collectBounds(follower, current, visited, min, max);
+			return bounds;
+		}
+		
+		private static final int DFS_VISITED = 1;
+		private static final int DFS_ON_STACK = 2;
+		
+		/**
+		 * Finding cycles is done with a depth-first search.
+		 */
+		private void findCycles(ISemState node, Map<ISemState, Set<ISemState>> cycles, Map<ISemState, Integer> dfsMark,
```

Comment:
This method could be generiefied and moved to org.eclipse.xtext.util.formallang.NfaUtil